File: uvaonlinejudge/11475 - Extend to Palindromes.c
#include <stdio.h>
#include <string.h>
#define MAX_SIZE 100000
#define BUFFER_SIZE MAX_SIZE * 2 + 3

char buffer[BUFFER_SIZE];
int prefix_func[MAX_SIZE * 2 + 1];

int read_test_case(char buffer[]){
    if(fgets(buffer,MAX_SIZE + 2, stdin) == NULL)
        return 0;
    return strlen(buffer) - 1;
}

void kmp_prefix_function(int n, char str[], int prefix_func[]){
    int i, j = 0;
    prefix_func[0] = 0;
    for(i = 1; i < n; i++){
        while(j > 0 && str[i] != str[j])
            j = prefix_func[j - 1];
        if(str[i] == str[j])
            j++;
        else
            j = 0;
        prefix_func[i] = j;
    }
}

int main(){
    int n, i;
    char *str = buffer + MAX_SIZE + 1;
    while((n = read_test_case(str)) > 0){
        for(i = 0; i < n; i++){
            buffer[MAX_SIZE - 1 - i] = str[i];
        }
        kmp_prefix_function(n * 2 + 1, buffer + MAX_SIZE - n , prefix_func);
        for(i = 0; i < n; i++)
            printf("%c", str[i]);
        for(i = n - prefix_func[n * 2] - 1; i >= 0; i--)
            printf("%c", str[i]);
        printf("\n");
    }
    return 0;
}

File: uvaonlinejudge/11475 - Extend to Palindromes.py
import sys

def read_test_case():
    line = sys.stdin.readline().strip()
    if len(line) == 0:
        return None
    return line

def kmp_prefix_function(string):
    n = len(string)
    prefix_func = [0] * n
    j = 0 
    for i in range(1, n):
        while j > 0 and string[i] != string[j]:
            j = prefix_func[j - 1]
        if string[i] == string[j]:
            j += 1
        else:
            j = 0
        prefix_func[i] = j
    return prefix_func

if __name__ == '__main__':
    for string in iter(read_test_case, None):
        n = len(string)
        prefix_func = kmp_prefix_function(''.join((string[::-1], '$', string)))
        print(string, string[:n - prefix_func[2 * n]][::-1], sep='')
